packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+    const streamName = Equinox.StreamName.create(this.categoryName, streamId)
+    const span = trace.getActiveSpan()
+    span?.setAttributes({
+      [Tags.access_strategy]: this.access.type,
+      [Tags.category]: this.categoryName,
+      [Tags.stream_name]: streamName,
+    })
+    switch (this.access.type) {
+      case "Unoptimized":
+        return this.context.loadBatched(streamName, this.codec.decode, this.fold, this.initial)
+      case "LatestKnownEvent":
+        return this.context.loadLast(streamName, this.codec.decode, this.fold, this.initial)
+      case "Snapshot":
+        return this.context.loadSnapshot(streamName, this.codec.decode, this.fold, this.initial, this.access.isOrigin)
+      case "RollingState":
+        return this.context.loadState(streamName, this.codec.decode, this.fold, this.initial)
+    }
+  }
+
+  async load(streamId: Equinox.StreamId, _maxStaleMs: number) {
+    const [token, state] = await this.loadAlgorithm(streamId)
+    return { token, state }
+  }
+
+  async reload(streamId: Equinox.StreamId, _requireLeader: boolean, t: TokenAndState<State>) {
+    if (this.access.type === "RollingState") {
+      return this.load(streamId, 0)
+    }
+    const streamName = Equinox.StreamName.create(this.categoryName, streamId)
+    const [token, state] = await this.context.reload(
+      streamName,
+      t.token,
+      this.codec.decode,
+      this.fold,
+      t.state,
+    )
+    return { token, state }
+  }
+
+  private updateSnapshot(
+    category: string,
+    streamName: string,
+    ctx: Context,
+    newState: State,
+    version: bigint,
+  ) {
+    if (this.access.type !== "Snapshot") return
+    const toSnapshot = this.access.toSnapshot
+    return async (trx: Transaction) => {
+      const snapshot = this.codec.encode(toSnapshot(newState), ctx)
+      const id = randomUUID()
+      const etag = randomUUID()
+      await trx.execute({
+        sql: `
+        INSERT INTO snapshots (stream_name, category, type, data, position, etag, id)
+        VALUES (?, ?, ?, ?, ?, ?, ?)
+        ON CONFLICT (stream_name) DO UPDATE
+        SET data = excluded.data, position = excluded.position, time = CURRENT_TIMESTAMP, type = excluded.type, id = excluded.id, etag = excluded.etag
+      `,
+        args: [streamName, category, snapshot.type, snapshot.data ?? null, version, etag, id],
+      })
+    }
+  }
+
+  async sync(
+    streamId: Equinox.StreamId,
+    ctx: Context,
+    token: StreamToken,
+    state: State,
+    events: Event[],
+  ): Promise<SyncResult<State>> {
+    const span = trace.getActiveSpan()
+    const streamName = Equinox.StreamName.create(this.categoryName, streamId)
+    span?.setAttributes({
+      [Tags.category]: this.categoryName,
+      [Tags.stream_name]: streamName,
+    })
+    if (this.access.type === "RollingState") {
+      if (events.length === 0) return { type: "Written", data: { token, state } }
```

Comment:
how is it not a hard programmer error if we come in here with no events?

Reply:
I mean, it'll get short circuited in `transactAsync`, but I can't write the type as a `NonEmptyArray<T>` so better to just handle the case? Or would you want a throw here?

Reply:
would go with a throw and a comment that it's for the compiler